### PR TITLE
Add expected type in incompatible arg errors

### LINKED ISSUE #3499: If list item has incompatible type, display expected type
Currently this program produces the error `List item 0 has incompatible type "float"`:

```py
from typing import List
def f(x: List[int]) -> None: ...
f([1.0])  # Error
```

This can be confusing, especially if the types are more complicated. The message doesn't tell what is expected, and the function definition can be hard to find, for example if it's in a stub. A better message would be something like this:

```
prog.py:3: error: List item 0 has incompatible type "float" (expected "int")
```

The same issue may also applies to dictionaries:

```
prog.py:3: error: Dict entry 0 has incompatible type "int": "int"
```

### PR DESCRIPTION
**Pull request by miedzinski**
Fixes #3499.

### CODE REVIEW

**Comment by miedzinski** [comment]
There is one test failing because we are using `format_simple` everywhere. Should we skip expected types when dealing with `FunctionLike` or use `format`?

**Comment by gvanrossum** [comment]
I'd start by not adding the "; expected blah" part to the error if format_simple() returns something empty.

**Comment by miedzinski** [comment]
Do we really want to do that? Even without this patch mypy prints malformed messages, e.g.

```python
(mypy-dev) ~/src/mypy [master] λ cat x.py 
from typing import Dict, Callable
def f() -> str:
    return 'foo'
d: Dict[str, Callable[..., int]] = {'bar': f}
(mypy-dev) ~/src/mypy [master] λ mypy x.py 
x.py:4: error: Dict entry 0 has incompatible type "str":
```

Notice the `:` at the end. And this doesn't happen with `Callable`s only (which I understand may get long), but even with `Type`.

I'd go for using `format` method, just like `MessageBuilder.invalid_index_type` does.

**Comment by JukkaL** [comment]
Using `format` seems reasonable to me.

File: mypy/messages.py [hunk 1]
```diff
@@ -536,29 +536,59 @@ def incompatible_argument(self, n: int, m: int, callee: CallableType, arg_type:
...
             name = callee.name[1:-1]
             n -= 1
             key_type, value_type = cast(TupleType, arg_type).items
-            msg = '{} entry {} has incompatible type {}: {}'.format(
-                name.title(), n, self.format_simple(key_type), self.format_simple(value_type))
+            expected_key_type, expected_value_type = cast(TupleType, callee.arg_types[0]).items
+
+            # don't increase verbosity unless there is need to do so
+            from mypy.subtypes import is_subtype
+            if is_subtype(key_type, expected_key_type):
```
Comment (ilevkivskyi) [inline mypy/messages.py]: ``Dict`` is invariant both in key and value types. Therefore ``is_subtype`` is probably not what you want here, maybe ``is_same_type`` from ``sametypes.py``?

File: mypy/messages.py [hunk 2]
```diff
@@ -536,29 +536,59 @@ def incompatible_argument(self, n: int, m: int, callee: CallableType, arg_type:
...
+                expected_key_type_str = self.format(expected_key_type)
+            else:
+                key_type_str, expected_key_type_str = self.format_distinctly(
+                    key_type, expected_key_type)
+            if is_subtype(value_type, expected_value_type):
+                value_type_str = self.format(value_type)
+                expected_value_type_str = self.format(expected_value_type)
+            else:
+                value_type_str, expected_value_type_str = self.format_distinctly(
+                    value_type, expected_value_type)
```
Comment (ilevkivskyi) [inline mypy/messages.py]: I would add few new tests checking that this verbosity logic is followed correctly in different scenarios (taking into account mentioned invariance).

**Comment by ilevkivskyi** [comment]
@miedzinski After some thinking, it looks like your initial design was actually right and it is OK to use ``is_subtype`` in the context of _inference_ for dict literals. Also the added tests don't trigger the new error message (this is related). So that I would restore the ``is_subtype`` and rewrite the added tests like you proposed on gitter.

Sorry for troubles.

**Pull request merged**